File: src/commands/others/ai/SVD.js
const exec = require("child_process").exec;
const path = require("node:path");
const fs = require("node:fs");

module.exports = SVD = async (interaction) => {
  const url = interaction.options.getAttachment("image").proxyURL;
  const value = interaction.options.getInteger("value").toString();
  fs.writeFileSync(
    "./src/commands/others/ai/resources/API.txt",
    url + "\n" + value
  );

  const pyExec = path.join(__dirname, "/python/SVD.py");
  exec(`python3 ${pyExec}`);

  await interaction.deferReply();

  let check = "";
  (function loop() {
    setTimeout(() => {
      check = fs.readFileSync(
        `${path.join(__dirname, "./resources/API.txt")}`,
        "utf-8"
      );
      if (check !== "done") {
        loop();
      } else {
        interaction.editReply({
          content: "",
          files: ["./src/commands/others/ai/resources/SVD.png"],
        });
      }
    }, 1000);
  })();
};

File: src/commands/others/ai/python/SVD.py
import numpy as np
from PIL import Image
import requests
from io import BytesIO
import time

path = "./src/commands/others/ai/resources/"

with open(path + "API.txt", "r") as f:
    part = f.readlines()

file = open(path + "API.txt", "w")

response = requests.get(part[0].strip())
img_gray = Image.open(BytesIO(response.content)).convert('LA')


value = int(part[1])

imgmat = np.array(list(img_gray.getdata(band=0)), float)
imgmat.shape = (img_gray.size[1], img_gray.size[0])
imgmat = np.matrix(imgmat)

U, sigma, V = np.linalg.svd(imgmat)

reconstimg = np.matrix(U[:, :value]) * \
    np.diag(sigma[:value]) * np.matrix(V[:value, :])

data = Image.fromarray(reconstimg).convert("L")

data.save(path + "SVD.png")
time.sleep(3)
file.write("done")
file.close
